src/enclave/Enclave/EnclaveContext.h

```diff
@@ -0,0 +1,215 @@
+#include "Enclave_t.h"
+#include <unordered_set>
+#include <vector>
+#include <string>
+#include <unordered_map>
+#include "../Common/common.h"
+#include "../Common/mCrypto.h"
+
+struct LogEntry;
+
+struct LogEntry {
+  std::string ecall;
+  int snd_pid;
+  int rcv_pid;
+  int job_id;
+
+  bool operator==(const LogEntry& le) const
+  { 
+      return (this->ecall == le.ecall && this->snd_pid == le.snd_pid && this->rcv_pid == le.rcv_pid && this->job_id == le.job_id); 
+  }
+}; 
+
+class LogEntryHashFunction { 
+public: 
+    // Example taken from https://www.geeksforgeeks.org/how-to-create-an-unordered_set-of-user-defined-class-or-struct-in-c/ 
+    size_t operator()(const LogEntry& le) const
+    { 
+        return (std::hash<std::string>()(le.ecall)) ^ (std::hash<int>()(le.snd_pid)) ^ (std::hash<int>()(le.rcv_pid)) ^ (std::hash<int>()(le.job_id)); 
+    } 
+};
+
+static Crypto mcrypto;
+
+class EnclaveContext {
+  private:
+    std::unordered_set<LogEntry, LogEntryHashFunction> ecall_log_entries;
+    int operators_ctr;
+    unsigned char shared_key[SGX_AESGCM_KEY_SIZE] = {0};
+
+    // For this ecall log entry
+    std::string this_ecall;
+    std::vector<std::vector<uint8_t>> log_entry_mac_lst;
+
+    std::string curr_row_writer;
+    // Special vectors of nonObliviousAggregateStep1
+    std::vector<std::vector<uint8_t>> first_row_log_entry_mac_lst;
+    std::vector<std::vector<uint8_t>> last_group_log_entry_mac_lst;
+    std::vector<std::vector<uint8_t>> last_row_log_entry_mac_lst;
+
+    int pid;
+    bool append_mac;
+
+    // Map of job ID for partition
+    std::unordered_map<int, int> pid_jobid;
```

Comment:
The `job_id` is incremented per partition after the partition finishes executing the ecall. Something like your scenario would likely render the replay attack defense idea (currently unimplemented but the idea was to use this `job_id` for that) infeasible. The scenario will likely also cause the JobVerificationEngine to output the wrong result -- right now it expects that every partition executes the exact same ordering and number of ecalls. We can address this in a future PR once this is merged to avoid adding even more to this one 😅.